internal/users/manager.go

```diff
@@ -115,55 +120,100 @@ func (m *Manager) UpdateUser(u types.UserInfo) (err error) {
 
 	log.Debugf(context.TODO(), "Updating user %q", u.Name)
 
+	if u.Name == "" {
+		return errors.New("empty username")
+	}
+
 	// authd uses lowercase usernames
 	u.Name = strings.ToLower(u.Name)
 
-	if u.Name == "" {
-		return errors.New("empty username")
+	// Prepend the user private group
+	u.Groups = append([]types.GroupInfo{{Name: u.Name, UGID: u.Name}}, u.Groups...)
+	userPrivateGroup := &u.Groups[0]
+
+	var currentUser *types.UserInfo
+	checkEqualUserExists := func() (exists bool, err error) {
+		// Check if the user already exists in the database.
+		currentUser, err = m.getCurrentUserFromDB(u.Name)
+		if err != nil {
+			return false, err
+		}
+		if currentUser == nil {
+			return false, nil
+		}
+		if !compareNewUserWithUserFromDB(u, *currentUser) {
+			log.Debugf(context.TODO(), "User %q is already in our database", u.Name)
+			return false, nil
+		}
+
+		log.Debugf(context.TODO(), "User %q in database already matches current", u.Name)
+		return true, nil
 	}
 
-	var uid uint32
+	// Do a first check before locking, so that if the user is already there and
+	// matches the DB entry, we can avoid any kind of locking (and so being
+	// blocked by other pre-auth users that may try to login meanwhile).
+	if exists, err := checkEqualUserExists(); exists || err != nil {
+		return err
+	}
 
-	// Prevent a TOCTOU race condition between the check for existence in our database and the registration of the
-	// temporary user/group records. This does not prevent a race condition where a user is created by some other NSS
-	// source, but that is handled in the temporaryRecords.RegisterUser and temporaryRecords.RegisterGroup functions.
-	m.updateUserMu.Lock()
-	defer m.updateUserMu.Unlock()
+	m.userManagementMu.Lock()
+	defer m.userManagementMu.Unlock()
 
-	// Check if the user already exists in the database
-	oldUser, err := m.db.UserByName(u.Name)
-	if err != nil && !errors.Is(err, db.NoDataFoundError{}) {
-		return fmt.Errorf("could not get user %q: %w", u.Name, err)
+	// Now that we're locked, check again if meanwhile some other request
+	// created the same user, if not we can do all the kinds of locking since
+	// we're sure that the user needs to be added or updated in the database.
+	if exists, err := checkEqualUserExists(); exists || err != nil {
+		return err
 	}
-	if errors.Is(err, db.NoDataFoundError{}) {
-		// Check if the user exists on the system
-		existingUser, err := user.Lookup(u.Name)
-		var unknownUserErr user.UnknownUserError
-		if !errors.As(err, &unknownUserErr) {
-			log.Errorf(context.Background(), "User already exists on the system: %+v", existingUser)
-			return fmt.Errorf("user %q already exists on the system (but not in this authd instance)", u.Name)
+
+	log.Debugf(context.TODO(), "User %q needs update", u.Name)
+
+	ctx, unlockEntries, err := localentries.ContextUserDBLocked(context.Background())
+	if err != nil {
+		return err
+	}
+	defer func() { err = errors.Join(err, unlockEntries()) }()
```

Comment:
As I mentioned before, I don't think it's a good idea to pass the whole local entries cache via the context, because contexts are intended for small, read-only values, and we're hiding too much behind the opaque context if we pass the cache with it.

Do you think it would be a lot of effort to instead pass a reference to a `EntriesWithLock` struct to the functions which need it?

Reply:
Well... Wasn't it similar to what we were doing before of 157a650ec?

Reply:
Huff... Hit Ctrl+Enter too quickly, however I can probably get back to the specific type, the context played well here I think though since it was hiding some details, and the data it stores is really small (well actually it's a read-only pointer, but even the pointed data only contains the refcounting and test information), so I kinda liked the idea of abstracting it.